xcdat/bounds.py

```diff
@@ -428,239 +468,319 @@ def _create_bounds(self, axis: CFAxisKey, coord_var: xr.DataArray) -> xr.DataArr
-    for step in time.values:
-        # If end of time interval and first day of year then subtract one month
-        # so bounds will be calculated correctly.
-        if (end_of_month) & (step.day < 2):
-            step = _month_add(step, -1, calendar)
+            year, month = step.year, step.month
 
-        year, month = step.year, step.month
+            l_bnd = obj_type(year, month, 1, 0, 0)
+            u_bnd = self._add_months_to_timestep(l_bnd, obj_type, delta=1)
 
-        l_bnd = cf_obj(year, month, 1, 0, 0)
-        u_bnd = _month_add(l_bnd, 1, calendar)
+            time_bnds.append([l_bnd, u_bnd])
 
-        time_bnds.append([l_bnd, u_bnd])
+        return time_bnds
 
-    da_time_bnds = xr.DataArray(
-        name=f"{time.name}_bnds",
-        data=time_bnds,
-        coords=dict(time=time),
-        dims=[*time.dims, "bnds"],
-        attrs={"xcdat_bounds": "True"},
-    )
+    def _add_months_to_timestep(
+        self,
+        timestep: Union[cftime.datetime, pd.Timestamp],
+        obj_type: Union[cftime.datetime, pd.Timestamp],
+        delta: int,
+    ) -> Union[cftime.datetime, pd.Timestamp]:
+        """Adds delta month(s) to a timestep.
 
-    return da_time_bnds
+        The delta value can be positive or negative (for subtraction). Refer to [1]_
+        for logic.
 
+        Parameters
+        ----------
+        timesep : Union[cftime.datime, pd.Timestamp]
+            A timestep represented as ``cftime.datetime`` or ``pd.Timestamp``.
+        obj_type : Union[cftime.datetime, pd.Timestamp]
+                The object type for time bounds based on the dtype of
+                ``timestep``.
+        delta : int
+            Integer months to be added to times (can be positive or negative)
 
-def create_daily_time_bounds(
-    time: xr.DataArray, freq: Literal[1, 2, 3, 4, 6, 8, 12, 24] = 1
-) -> xr.DataArray:
-    """Creates time bounds for each timestep with the start and end of the day.
+        Returns
+        -------
+        Union[cftime.datetime, pd.Timestamp]
 
-    Bounds for each timestep corresponds to 00:00:00 timepoint on the
-    current day and 00:00:00 on the subsequent day.
+        References
+        ----------
+        [1] https://stackoverflow.com/a/4131114
+        """
+        # Compute the new month and year with the delta month(s).
+        month = timestep.month - 1 + delta
+        year = timestep.year + month // 12
+        month = month % 12 + 1
+
+        # Re-use existing hour/minute/second/microsecond.
+        hour = timestep.hour
+        minute = timestep.minute
+        second = timestep.second
+        microsecond = timestep.microsecond
+
+        # If day is greater than days in month use days in month as day.
+        day = timestep.day
+        dim = obj_type(year, month, 1).daysinmonth
+        day = min(day, dim)
+
+        # Create the new timestep.
+        new_timestep = obj_type(year, month, day, hour, minute, second, microsecond)
+
+        return new_timestep
```

Comment:
`_add_months_to_timestep` method is based on `_month_add()` and replaces `_month_add()`. 

I added this method to `BoundsAccessor` because it is a private method that is only called in `_create_monthly_time_bounds()`. I removed logic related to arrays because we only pass a single timestep (`cftime.datetime` or `pd.Timestamp`) internally.